[Draft] CollisionMonitor: add CostmapSource + dataset-based bag test

Lotusymt opened the pull request:
<!-- Please fill out the following pull request template for non-trivial changes to help us process your PR faster and more efficiently.-->

---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | Fixes/Implements #4794 |
| Primary OS tested on | Ubuntu 22.04 (ROS 2 Rolling) |
| Robotic platform tested on | unit tests + launch test (no physical robot) |
| Does this PR contain AI generated software? | Yes and it is marked inline in the code |
| Was this PR description generated by AI software? | No |

---

## Description of contribution in a few bullet points

* Add **`CostmapSource`** to Collision Monitor to allow collision checks against a subscribed **local costmap** (`nav2_msgs/Costmap`).
* Wire `CostmapSource` into `CollisionMonitor::configureSources()` and expose params:
  * `*.topic`, `*.cost_threshold` (0–255), `*.treat_unknown_as_obstacle` (bool).
* Provide minimal example in `params/collision_monitor_params.yaml`.
* Update `README.md`:
  * Add **Costmap** to the list of supported sources.
  * Add a concise note about **latency warning** when using a costmap source.
* Add **dataset-based integration test** (bag replay):
  * `test/collision_monitor_node_bag.cpp` (metrics: time-to-stop, hold-stop%, time-to-resume, false-stop%).
  * `test/collision_monitor_node_bag.launch.py` to bring up CM, play a tiny bag, and run gtest.
  * Tiny **mcap.zstd** bag + test YAML installed for `get_package_share_directory()`.


---

## Description of documentation updates required from your changes

* **docs.nav2.org** (separate PR): add a short `.. warning::` box on the Collision Monitor page describing the **trade-offs** of using a costmap source (persistence vs. latency/staleness), and list the new parameters.
* Ensure the new source and parameters are included in any **parameter reference** and **tuning** sections.
* (This PR) `README.md` already updated with a brief note and source entry.

---

## Description of how this change was tested

- **Unit/launch tests:** Added a launch gtest that replays a tiny bag and checks CM behavior (time-to-stop, hold-stop%, time-to-resume, false-stop%). Ran `colcon test --packages-select nav2_collision_monitor`; existing unit tests + the new launch test pass on Ubuntu.
- **ABI:** Verified that ABI was preserved.


---

## Future work that may be required in bullet points


#### For Maintainers: <!-- DO NOT EDIT OR REMOVE -->
- [ ] Check that any new parameters added are updated in docs.nav2.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists
- [ ] Should this be backported to current distributions? If so, tag with `backport-*`.

SteveMacenski commented on nav2_collision_monitor/package.xml:
```diff
@@ -27,11 +27,15 @@
   <depend>nav2_ros_common</depend>
   <depend>point_cloud_transport</depend>
   <depend>point_cloud_transport_plugins</depend>
+  <!-- <depend>nav2_common</depend> -->
```
remove?

SteveMacenski commented on nav2_collision_monitor/include/nav2_collision_monitor/costmap.hpp:
```diff
@@ -0,0 +1,72 @@
+{
+public:
+  CostmapSource(
+
```
Remove extra sapce

SteveMacenski commented on nav2_collision_monitor/include/nav2_collision_monitor/costmap.hpp:
```diff
@@ -0,0 +1,72 @@
+  // ↑ Store the latest Costmap message; we’ll read it in getData()
+
+  nav2_msgs::msg::Costmap::ConstSharedPtr data_;     // Latest costmap message (thread-safe shared ptr)
+  rclcpp::Subscription<nav2_msgs::msg::Costmap>::SharedPtr data_sub_;
```
Use `nav2::` subscriptions instead

SteveMacenski commented on nav2_collision_monitor/src/costmap.cpp:
```diff
@@ -0,0 +1,149 @@
+  return true;
+}
+
+void CostmapSource::getParameters(std::string & source_topic)
```
All params should be added to  the collision monitor's configuration guide at docs.nav2.org

SteveMacenski commented on nav2_collision_monitor/src/costmap.cpp:
```diff
@@ -0,0 +1,149 @@
+// Copyright (c) 2025 Angsa Robotics
```
There's alot of weird spacing issues in this file (too many at times, none at others, etc). Please review. Examples
L27, L63, L81, L91L100, ...

SteveMacenski commented on nav2_collision_monitor/src/costmap.cpp:
```diff
@@ -0,0 +1,149 @@
+  const auto thresh_name = source_name_ + ".cost_threshold";
+  nav2::declare_parameter_if_not_declared(
+      node, thresh_name, rclcpp::ParameterValue(253));
+  int v = node->get_parameter(thresh_name).as_int();
```
Use the new declare_or_get_parameter API in nav2::LIfecycleNode

SteveMacenski commented on nav2_collision_monitor/test/bags/cm_moving_obstacle/fake_cm_bag_source.py:
```diff
@@ -0,0 +1,115 @@
+    def __init__(self):
+        super().__init__('fake_cm_bag_source')
+        # Use sim time so /clock drives timestamps
+        # self.set_parameters([Parameter('use_sim_time', value=True)])
```
Commented out?

SteveMacenski commented on nav2_collision_monitor/test/collision_monitor_node_bag.cpp:
```diff
@@ -0,0 +1,241 @@
+    // Window where obstacle exists in the bag
+    stop_window_start_ = this->declare_parameter<double>("stop_window_start", 3.0);
+    stop_window_end_ = this->declare_parameter<double>("stop_window_end", 8.0);
+
```
```suggestion
```

SteveMacenski reviewed (commented):
Checkout Linting / CI jobs that are failing. You may need to pull in main / rebase as well. 

Please provide some comments inline in the test files so someone can understand what each are doing -- its a little difficult to read them and understand (1) fully what each do, (2) the difference in what one covers the other doesn't, and (3) the intent 

But, by in large, this looks great!

Lotusymt commented:
Thanks for the detailed review, Steve!
I’ll address the issues over the next couple of days and push updates soon.

Lotusymt commented on nav2_collision_monitor/test/collision_monitor_node_test.cpp:
```diff
@@ -945,6 +945,7 @@ TEST_F(Tester, testPolygonSource)
   sendTransforms(curr_time);
 
   // 1. Obstacle is far away from robot
+  curr_time = cm_->now();
```
I think the original test will still work on most machines, but it failed on mine, probably because the machine was a bit slow and the message timestamp became stale. Updating current_time right before sending fixes that and makes the test more robust.

SteveMacenski commented on nav2_collision_monitor/src/costmap.cpp:
```diff
@@ -0,0 +1,146 @@
+
+namespace nav2_collision_monitor
+{
+CostmapSource::CostmapSource(
```
```suggestion

CostmapSource::CostmapSource(
```

SteveMacenski commented on nav2_collision_monitor/src/costmap.cpp:
```diff
@@ -0,0 +1,146 @@
+  std::string source_topic;
+  getParameters(source_topic);
+  data_sub_ = node->create_subscription<nav2_msgs::msg::Costmap>(
+      source_topic,
```
4 indents? Should be 2

SteveMacenski commented on nav2_collision_monitor/test/bags/cm_moving_obstacle/fake_cm_bag_source.py:
```diff
@@ -0,0 +1,117 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
```
Misformatted?

Lotusymt commented:
Hi Steve,

I’ve updated the PR to address the issues you pointed out:

- Removed the coverage exclusion and the old comment.
- Cleaned up the formatting you mentioned earlier.

Right now, the uncovered lines are mainly the branch that handles malformed `/local_costmap/costmap` data, which should not trigger in the normal “happy path”.

If you think we should explicitly cover this failure mode, I’m happy to add a small unit test that publishes invalid data and hits this branch. Please let me know your preference, and I can update the PR accordingly. Thank you!

SteveMacenski commented:
> If you think we should explicitly cover this failure mode, I’m happy to add a small unit test that publishes invalid data and hits this branch

That would be nice, yes. It should be easy by sending it a bogus costmap manually in the unit tests.

Also, do you think the costmap source should be added to the collision detector node as well? 

Finally, just need to update the migration guide in https://github.com/ros-navigation/docs.nav2.org/pull/804 and we should be ready to merge!

SteveMacenski commented:
Got some errors here in the build:

```
--- stderr: nav2_collision_monitor                               
ld.lld: error: undefined reference due to --no-allow-shlib-undefined: nav2_collision_monitor::CostmapSource::CostmapSource(std::weak_ptr<nav2::LifecycleNode> const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>> const&, std::shared_ptr<tf2_ros::Buffer>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>> const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>> const&, std::chrono::duration<long, std::ratio<1l, 1000000000l>> const&, rclcpp::Duration const&, bool)
>>> referenced by libcollision_detector_core.so

ld.lld: error: undefined reference due to --no-allow-shlib-undefined: nav2_collision_monitor::CostmapSource::configure()
>>> referenced by libcollision_detector_core.so
collect2: error: ld returned 1 exit status
gmake[2]: *** [CMakeFiles/collision_detector.dir/build.make:291: collision_detector] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:269: CMakeFiles/collision_detector.dir/all] Error 2
gmake: *** [Makefile:146: all] Error 2
---
```

Once that is resolved I approve the pair of PRs :-)

mini-1235 commented on nav2_collision_monitor/launch/collision_detector_node.launch.py:
```diff
@@ -99,7 +99,8 @@ def generate_launch_description() -> LaunchDescription:
                 output='screen',
                 emulate_tty=True,  # https://github.com/ros2/launch/issues/188
-                parameters=[{'autostart': autostart}, {'node_names': lifecycle_nodes}],
+                parameters=[{'autostart': autostart},
```
Same here

Lotusymt commented on nav2_collision_monitor/test/CMakeLists.txt:
```diff
@@ -52,9 +54,38 @@ ament_add_gtest(collision_detector_node_test collision_detector_node_test.cpp)
+install(FILES collision_monitor_node_bag.yaml
+  DESTINATION share/${PROJECT_NAME}/test)
+
+find_package(GTest REQUIRED)
```
Thanks for pointing this out!
I kept this as an add_executable on purpose: the binary is a gtest-based helper that’s launched from collision_monitor_node_bag.launch.py.
If we register it with ament_add_gtest, colcon test / CTest would try to run it standalone without that launch setup, so it would block waiting for /clock and /cmd_vel or fail with no data.

Lotusymt replied:
Good point, thank you for pointing that out! Your suggestion would definitely be more consistent with the existing style. Since this PR is already merged, I’ll keep the current version for now, but I’ll keep this pattern in mind for future tests.

Lotusymt commented on nav2_collision_monitor/launch/collision_monitor_node.launch.py:
```diff
@@ -11,7 +11,7 @@
 # distributed under the License is distributed on an "AS IS" BASIS,
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-# See the License for the specific language governing permissions and
+# See the License for the specific languazge governing permissions and
```
Thanks for catching that — I totally agree this change is unrelated to the PR.
I think it was accidentally modified earlier when I was experimenting with a small script to debug some linter issues.
I’ve now reverted the line back to the original license text, and the linter tests still pass.

arjangupta commented:
Hi all,

I'm wondering if this feature will get backported to ROS Jazzy?

Thanks!
Arjan
